File: js/077_Combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    var ret = [];
    var path = [];
    backtrack(ret,path,1,n,k)
    return ret;
};

var backtrack = function(ret,path,index,n,k){
    if(path.length == k){
        ret.push(path.concat());
        return;
    }
    for(var i = index; i <= n; ++i){
        path.push(i);
        backtrack(ret,path,i + 1, n ,k);
        path.pop();
    }
}

File: cpp/077_Combinations.cpp
//two recursive
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ret;
        vector<int> path;
        backtrack(ret,path,1,n,k);
        return ret;
    }
private:
    void backtrack(vector<vector<int>>& ret, vector<int> path, int m, int n, int k){
        if(path.size() == k){
            ret.push_back(path);
            return;
        }
        if(m > n) return;
        path.push_back(m);
        backtrack(ret,path,m+1,n,k);
        path.pop_back();
        backtrack(ret,path,m+1,n,k);
    }
};

//one recursive
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ret;
        vector<int> path;
        backtrack(ret,path,1,n,k);
        return ret;
        
    }
private:
    void backtrack(vector<vector<int>>& ret, vector<int>& path, int index,int n, int k){
        if(path.size() == k){
            ret.push_back(path);
            return;
        }
        for(int i = index; i <= n; ++i){
            path.push_back(i);
            backtrack(ret,path,i+1,n,k);
            path.pop_back();
        }
    }
};